games/guess-it/src/lib.rs

```diff
@@ -0,0 +1,288 @@
+
+            let key = auth.get_non_fungible_key();
+            let player = self.game.players.get_mut(&key).unwrap();
+            player.make_guess(guess.clone());
+
+            let guesses: Vec<String> = self.game.players.clone().into_iter()
+                .map(|(_key, player)| player.get_guess())
+                .filter(|guess| !guess.is_empty())
+                .collect();
+            self.game.state = if guesses.len() == 2 { State::WinnerSelection } else { State::MakeGuess };
+            if self.game.state == State::WinnerSelection {
+                let response = self.check_winner();
+                if self.game.winner != NonFungibleKey::from(0).to_string() {
+                    self.game.state = State::Payout;
+                }
+                format!("Your guess '{}' was accepted\nResponse: {}", guess, response)
+            } else {
+                format!("Your guess '{}' was accepted", guess)
+            }
+
+        }
+
+        pub fn check_winner(&mut self) -> String {
+            // Pick random number and store it
+            let random_number = (get_random(&self.game) % 6) + 1;
+            // determine who was closest without going over
+            self.game.winner = get_winner(&self.game, random_number);
+            self.game.last_roll = random_number.to_string();
+            // update player's share of pot
+            format!("Dice roll was: '{}' .... winner is: {}",random_number, self.game.winner)
+        }
+
+        #[auth(badge_ref)]
+        pub fn withdraw_funds(&mut self) -> (Bucket, String) {
+            assert_eq!(self.game.state, State::Payout, "It's not time to get paid!");
+            let nft_id: NonFungibleKey = auth.get_non_fungible_key();
+
+            if nft_id.to_string() == self.game.winner {
+                let amt = self.bank.amount();
+                let payout = self.bank.take(amt);
+
+                (payout, format!("You withdrew {:?} XRD!", amt))
+            }
+            else { panic!("You cannot withdraw funds!") }
+        }
+
+        pub fn get_context(&self) -> (Actor, Address, H256, u64, u128) {
+            context::ContextTest::query()
+        }
+    }
+}
+
+fn get_random(game: &Game) -> u128 {
+    let multiplier = game.players.clone().into_iter()
+        .map(|(_,p)| u128::from_str(p.guess.as_str()).unwrap_or(0))
+        .reduce(|a,b| a * b).unwrap_or(1);
+
+    Uuid::generate() * multiplier
+}
+
+fn get_winner(game: &Game, random_number: u128) -> String {
+    let player_ids: Vec<NonFungibleKey> = game.players.clone().into_iter().map(|(k,_)| k).collect();
+    let p1 = random_number - u128::from_str(game.players.get(&player_ids[0]).unwrap().guess.as_str()).unwrap_or(0);
+    let p2 = random_number - u128::from_str(game.players.get(&player_ids[1]).unwrap().guess.as_str()).unwrap_or(0);
+    let closest_guess = cmp::min(p1, p2);
+    let winner = if p1 == closest_guess { &player_ids[0] } else { &player_ids[1] };
```

Comment:
In the case where both players guess the same number, `p1` would have an advantage over `p2` as they would always win in the case of a tie.

Reply:
I'll address this in the next commit